feat(video-player): add component (VIV-1486)

TaylorJ76 opened the pull request.

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+@use "../../../../shared/src/lib/sass/mixins/border-radius" as border-radius-variables;
+@import url("../../../../../node_modules/video.js/dist/video-js.css");
```
```suggestion
@import url("../../../../../node_modules/video.js/dist/video-js.css");
@use "../../../../../dist/libs/tokens/scss/tokens.constants" as constants;

```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+.vjs-theme-vivid { 
+	position: relative;
+	border-radius: #{border-radius-variables.$border-radius-normal};
+	font: var(--vvd-typography-base);
```
```suggestion
	font: var(#{constants.$vvd-typography-base});
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	font: var(--vvd-typography-base);
+	
+	.vjs-big-play-button {
+		width: 2.5em;
+		height: 2.5em;
+		margin-top: -1.25em;
+		margin-left: -1.25em;
```
```suggestion
                 $big-button-size: 2.5em;
		inline-size: $big-button-size;
		block-size: $big-button-size;
		margin-block-start: -$big-button-size;
		margin-inline-start -$big-button-size;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+
+	.vjs-time-control { 
+		display: block; 
+		width: auto;
+		min-width: auto; 
```
```suggestion
		inline-size: auto;
		min-inline-size: auto; 
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+		width: auto;
+		min-width: auto; 
+		padding: 0; 
+		line-height: 4em;
```
```suggestion
		line-height: 4;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	.vjs-control-bar {
+		width: calc(100% - 8px);
+		height: 4em;
```
```suggestion
		inline-size: calc(100% - 8px);
		block-size: 4em;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+
+		ul {
+			border-radius:
+				#{border-radius-variables.$border-radius-normal} 
```
same as before :)

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+
+	.vjs-menu-button {
+		border-radius: 
+			#{border-radius-variables.$border-radius-none} 
```
not code related - but why are not we keeping the border-radius to the top as well - we don't have this kind of style in vivivd even when a menu is opened

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	.vjs-menu {
+		left: -4em;
+		width: 12em;
```
```suggestion
		inset-inline-start: -4em;
		inline-size: 12em;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+		}
+
+		.vjs-playback-rate & {
+			left: 0;
+			width: 4em;
```
```suggestion
		inset-inline-start: 0;
		inline-size: 4em;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+		}
+
+		li:not(.vjs-selected, .vjs-menu-title):hover {
+			background: #000;
```
```suggestion
			background: var(#{constants.$vvd-color-canvas-text});
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	.vjs-menu-button-popup .vjs-menu {
+		margin-bottom: 2.5em;
```
```suggestion
		margin-block-end: 2.5em;
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	.vjs-button:hover {
+		background: #000;
```
```suggestion
		background: var(#{constants.$vvd-color-canvas-text});
```

rachelbt commented on libs/components/src/lib/video-player/video-player.scss:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	.vjs-volume-tooltip, .vjs-time-tooltip {
+		top: -5em;
```
```suggestion
		inset-block-start: -5em;
```

rachelbt commented on libs/components/src/lib/video-player/README.md:
````diff
@@ -0,0 +1,222 @@
+<script type="module">import '@vonage/vivid/video-player';</script>
+```
+
+## Usage
````
```suggestion
## Members
```

rachelbt commented on libs/components/src/lib/video-player/README.md:
```diff
@@ -0,0 +1,222 @@
+
+### Video source
+
+#### src attribute
```
```suggestion
#### src attribute

Type: string
Default: undefined
```
not sure this is the correct values - but this is missing for all the members

rachelbt reviewed (commented):
nice :)
Went over the docs and the css.
The doc page is not aligned to other pages, I commented in some places but not all (laziness :) )

I think that the looks of the video player is not fully "vivid" especially the big play button.

the controls of the video in mobile are not responsive and causing horizontal scroll:
![image](https://github.com/Vonage/vivid-3/assets/7512792/eec5bc17-39d0-4db3-93f3-2833ab0e12cd)

I left css comments and this is one in general - we are not using em in vivid, I see the need here to keep it responsive but it is a bit strange here, not sure what we need to do about it :)

YonatanKra reviewed (changes requested):
One big component. Nice work!
A few changes to cope with Vivid standards.
![image](https://github.com/Vonage/vivid-3/assets/6459899/08f0ff0f-a951-4ff8-ac4f-59327342d20b)

TaylorJ76 commented on libs/components/src/lib/enums.ts:
```diff
@@ -74,3 +74,10 @@ export enum AriaLive {
 	Assertive = 'assertive',
 }
+
+export enum SkipBy {
+	Zero = '0',
+	Five = '5',
+	Ten = '10',
+	Thirty = '30',
+}
```
The reason I didn't is that we have a request for audio player to have the same functionality.

RichardHelm commented on libs/components/src/locales/en-GB.ts:
```diff
@@ -103,6 +104,103 @@ const enGB: Locale = {
 	splitButton: {
 		showMoreActionsLabel: 'Show more actions',
+	},
+	videoPlayer: {
+		'Audio Player': 'Audio Player',
```
I think we can just import the locale from video.js to avoid copying it:
```suggestion
	videoPlayer: {
		videoJsLocale: <...imported from video.js...>,
```

TaylorJ76 commented on libs/components/src/lib/video-player/README.md:
````diff
@@ -0,0 +1,222 @@
+
+The poster image is displayed before the video has been interacted with. The `poster` attribute accepts a url string (absolute or relative) to an image.
+
+```html preview
+<style>
+	html {
+		max-width: 800px; /* for demo purposes */
+	}
+</style>
````
```html preview 800px - Makes this preview panel 800px height, which doesn't look great.

TaylorJ76 commented:
> nice :) Went over the docs and the css. The doc page is not aligned to other pages, I commented in some places but not all (laziness :) )
> 
> I think that the looks of the video player is not fully "vivid" especially the big play button.
> 
> the controls of the video in mobile are not responsive and causing horizontal scroll: ![image](https://private-user-images.githubusercontent.com/7512792/306909855-eec5bc17-39d0-4db3-93f3-2833ab0e12cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2ODUwOTksIm5iZiI6MTcwODY4NDc5OSwicGF0aCI6Ii83NTEyNzkyLzMwNjkwOTg1NS1lZWM1YmMxNy0zOWQwLTRkYjMtOTNmMy0yODMzYWIwZTEyY2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjNUMTAzOTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTY2YWE5MWUyYWUyYjFmODc3NTEzZTMzZjM5MzlhZGY5Y2VlODRkMDBkYmMyOTk3MDFjY2FkNTNkNDQyZTczYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dItCaiBQpGYlE-qb-UGgf1ILwP_90eAOdIv8fh2m0tk)
> 
> I left css comments and this is one in general - we are not using em in vivid, I see the need here to keep it responsive but it is a bit strange here, not sure what we need to do about it :)

I have sorted the responsiveness using container queries and converted `em`s to `px`s
And I have changed the big button to be smaller and hopefully more Vivid-like.

TaylorJ76 commented on libs/components/src/lib/video-player/IconFontStyles.ts:
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable max-len */
+export const iconFontStyles = `@font-face {
```
I found a way to delete the icon font stuff all together. Now using the built in svgs

TaylorJ76 commented on libs/components/src/lib/video-player/IconFontStyles.ts:
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable max-len */
+export const iconFontStyles = `@font-face {
+    font-family: VideoJS;
```
I have found a way to remove this completely.

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+		});
+	});
+
+	describe('src', () => {
```
Are the below `it`s all the possible use cases for changing `src`? For instance 
- What happens when you remove `src`?
- What are valid values for `src`?
- How do you handle invalid `src`?

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+	});
+
+	describe('autoplay', () => {
+		it('should set the muted attribute on the video element', async () => {
```
This description doesn't seem to be the case anymore. Can you set a more accurate description please?

```suggestion
		it('should set reflect autoplay on the video element', async () => {
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+		});
+	});
+
+	describe('autoplay', () => {
```
Again, it seems like `autoplay` should have more tests. For instance 
- Does removal works the same?
- Can this be used as a programatic API? (e.g. element.autoplay = true;)

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+			}
+		});
+
+		it('should disable playback rates when passed an empty string', async () => {
```
According to the test, it should hide the playback rates, right?
```suggestion
		it('should hide playback rates when passed an empty string', async () => {
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+	});
+
+	describe('captions', () => {
+		it('displays the captions button and menu', async () => {
```
```suggestion
		it('should display the captions button and menu when caption tracks are set', async () => {
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+						src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/captions.en.vtt" 
+						srclang="en" 
+						label="English" 
+						default>
```
I suggest not to use external content. This can actually be replaced with any other string (even an empty one) because what matters is the number of tracks of `kind="captions"`.
```suggestion
						src=""
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+						kind="captions" 
+						src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/captions.en.vtt" 
+						srclang="fr" 
+						label="French">
```
```suggestion
						src=""
```

TaylorJ76 replied:
Not sure why we need to remove the label

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelectorAll('[kind="captions"]').length).toBe(2);
```
This expectation means nothing. All it is saying: "if I add two tracks of kind captions, there should be two elements of kind captions". Where's the captions button test? And probably the captions selection options test?

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelectorAll('[kind="descriptions"]').length).toBe(2);
```
As above with captions - I cannot see how this expects a button. And as above, is the button all that's added? Or is there a menu to select from?

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelector('[kind="chapters"]')).not.toBe(null);
```
As above - what should really happen. In this case, you are using `not`. This should be avoided most of the times, as negative assertions are a code smell

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+	});
+
+	describe('skip-by buttons', () => {
+		it('should modify skip button amount', async () => {
+			element = (await fixture(
+				`<${COMPONENT_TAG} skip-by="30">
+					<source src="${VIDEO_SRC}" type="video/mp4">
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			const SkipBackwardBtn = element.shadowRoot?.querySelector('.vjs-skip-backward-30');
+			expect(SkipBackwardBtn?.classList.contains('vjs-hidden')).toBe(false);
+			const SkipForwardBtn = element.shadowRoot?.querySelector('.vjs-skip-forward-30');
+			expect(SkipForwardBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
```
This test is more "correct" IMO:
```suggestion
		it('should modify skip button amount', async () => {
			element = (await fixture(
				`<${COMPONENT_TAG} skip-by="30">
					<source src="${VIDEO_SRC}" type="video/mp4">
				</${COMPONENT_TAG}>`
			)) as VideoPlayer;
			const [skipBackwardButtonWhenValueIs30, skipForwardButtonWhenValueIs30] = getSkipButtons(30);
			
			element.skipBy = 15;
			const [skipBackwardButtonWhenValueIs15, skipForwardButtonWhenValueIs15] = getSkipButtons(15);
			
			expect(skipBackwardButtonWhenValueIs30).toBeTruthy();
			expect(skipForwardButtonWhenValueIs30).toBeTruthy();
			expect(skipBackwardButtonWhenValueIs15).toBeTruthy();
			expect(skipForwardButtonWhenValueIs15).toBeTruthy();
			expect(getSkipButtons(30)[0]).toBeNull();
			expect(getSkipButtons(30)[1]).toBeNull();
		});
```
Notice I've used some helper function to make the code easier to read (it needs to be implemented - and of course reused in the rest of the tests using skip buttons :) ).

Due to its length, I'd even split it to 3 tests:
1. Initial state when no skip-by is given
2. Initial state when skip-by is given
3. Initial state when skip-by is given and value is changed dynamically (via JS API)
This will create 3 short tests that explain the API

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+			expect(SkipForwardBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
+
+		it('should disable skip by buttons when passed 0', async () => {
```
That's a great `edge case` test!

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,286 @@
+
+	describe('a11y', () => {
+		// skipped because there are aria tags (menus) inside video.js that are not correct
+		xit('should pass html a11y test', async () => {
```
This is a skipped test. Either remove it or make it pass :)

YonatanKra reviewed (changes requested):
A few testing comments.

YonatanKra commented on libs/components/src/lib/video-player/video-player.template.ts:
```diff
@@ -0,0 +1,27 @@
+	context: ElementDefinitionContext,
+	definition: FoundationElementDefinition
+) => ViewTemplate<VideoPlayer> = (
+	// context: ElementDefinitionContext
```
This can be removed

YonatanKra commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,179 @@
+	/**
+	 * @internal
+	 */
+	initialiseVideo() {
```
Two things:
1. This is funny, actually. Now I understand why `init` functions are `init` functions - to keep the peace between UK and US devs ;) My laptop screams at `initialise` and wants to turn it into `initialize` ;)
2. Can this be a private function as well? (`#initVideo`)

YonatanKra commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,179 @@
+	/**
+	 * @internal
+	 */
+	getSettings() {
```
Can this method be private (`#getSettings`)? 

Or even better, a private getter? 
```
#settings() {
...
}
```
Right now it is exposed and part of our API.

YonatanKra commented on libs/components/tsconfig.json:
```diff
@@ -21,7 +21,8 @@
     "removeComments": true,
-    "composite": true
+    "composite": true,
+    "resolveJsonModule": true
```
Where do we import a `.json`?

TaylorJ76 replied:
In the locales files. We import the locales provided by video.js.

YonatanKra commented on libs/wrapper-gen/src/generator/polyfill/dom.ts:
```diff
@@ -23,4 +23,14 @@ global.matchMedia =
       removeListener: () => {},
     } as unknown as MediaQueryList));
 
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+global.addEventListener = () => {};
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+global.removeEventListener = () => {};
+
+Object.defineProperty(global.document, 'readyState', {
+	value: 'complete',
+});
+
```
What does this solve?

TaylorJ76 replied:
I think @RichardHelm added this as a fix to allow the docs site to deploy.

YonatanKra commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,179 @@
+	/**
+	 * @internal
+	 */
+	player: any;
```
This can probably be a private property (`#player`)

TaylorJ76 replied:
We are using it in the tests to mock the events.

YonatanKra commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,179 @@
+	/**
+	 * @internal
+	 */
+	videoEle: any;
```
Again, this should probably be private.

YonatanKra reviewed (changes requested):
None tests related comments

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,342 @@
+
+	describe('src', () => {
+		describe('no src provided', () => {
+			it('should show the internal invalid src error message', async() => {
```
```suggestion
			it('should show the invalid src error when no src provided', async() => {
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,342 @@
+	});
+
+	describe('src', () => {
+		describe('no src provided', () => {
```
can remove this describe as it holds only one usecase (see my suggestion for the `it`)

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,342 @@
+					`<${COMPONENT_TAG}></${COMPONENT_TAG}>`
+				)) as VideoPlayer;
+				await elementUpdated(element);
+				const noSrcErrorEl = element.shadowRoot!.getElementById('no-sources');
+				const dialogContentEl = getDialogContentEle();
+				expect(noSrcErrorEl?.classList.contains('vjs-hidden')).toBe(false);
+				expect(dialogContentEl!.textContent?.trim()).toBe('No compatible source was found for this media.');
```
This is a bit confusing for me.
Creating the component with no `src` means two elements change? What's the connection between these two elements?

TaylorJ76 replied:
When there is no src. Video.js appears as if it has loaded properly and displays the big play button.
I thought it would be more informative to the consumer if a message was displayed.

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,342 @@
+			});
+		});
+
+		it('should show the big play button by removing the vjs-hidden class', async () => {
```
```suggestion
		it('should remove the vjs-hidden class when src is set', async () => {
```

YonatanKra commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,342 @@
+			expect(element.playbackRates).toBe(DEFAULT_PLAYBACK_RATES);
+		});
+
+		it('should show the big play button by removing the vjs-hidden class', async () => {
```
This is part of the `src` tests. You have the exact same test there.
This part is supposed to be the `clean init test` - so maybe test here that hidden class is set on first init with no params.

YonatanKra commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,213 @@
+	#initVideo() {
+		if (this.player) this.player.dispose();
+		const settings = this.#getSettings();
+		const noSourcesError = this.shadowRoot!.getElementById('no-sources');
+
+		this.#videoEle = document.createElement('video');
+		const trackEles = this.querySelectorAll('track');
+		for(let x = 0; x < trackEles.length; x++) {
+			this.#videoEle.appendChild(trackEles[x]);
+		}
+		this.#videoEle.setAttribute('crossorigin', 'anonymous');
+		this.#videoEle.setAttribute('playsinline', '');
+		if (this.loop) this.#videoEle.setAttribute('loop', '');
+		if (this.autoplay) this.#videoEle.setAttribute('autoplay', '');
+		const control = this.shadowRoot!.querySelector('.control');
+
+		if (settings.sources && control && noSourcesError) {
+			noSourcesError.classList.add('vjs-hidden');
+			control.appendChild(this.#videoEle);
+			this.player = videojs(this.#videoEle, settings);
+			// removes lang="current" to avoid clash with vivid localization
+			this.shadowRoot!.querySelector('[lang]')!.removeAttribute('lang'); 
+			
+			this.player.on('play', () => this.$emit('play'));
+			this.player.on('pause', () => this.$emit('pause'));
+			this.player.on('ended', () => this.$emit('ended'));
+		} else {
+			if (noSourcesError) noSourcesError.classList.remove('vjs-hidden');
+		}
```
IMO most of it should go to the component's template. WDYT?

rachelbt reviewed (commented):
Docs look great! 
So much more informative and tidy! 👍

rachelbt reviewed (commented):
2 last ui test are nor rendering :(

YonatanKra commented on libs/components/src/lib/video-player/ui.test.ts:
```diff
@@ -0,0 +1,159 @@
+	const loadProgress = await page.$('#video-player-controls .vjs-load-progress');
+	const loadProgress2 = await page.$('#video-player-controls-2 .vjs-load-progress');
+	// hide elements that could lead to a flakey test
+	videoEle?.evaluate(element => element.style.visibility = 'hidden');
```
Don't we want to test user behavior? You can emulate user clicks and actual interaction with the video...
Can we use the above instead of using `evaluate`?

RichardHelm commented on libs/components/src/lib/video-player/locale.ts:
```diff
@@ -0,0 +1,97 @@
+	'Cyan'?: string;
+	'Background'?: string;
+	'Window'?: string;
+	'Transparent'?: string;
+	'Semi-Transparent'?: string;
+	'Opaque'?: string;
+	'Font Size'?: string;
+	'Text Edge Style'?: string;
+	'None'?: string;
+	'Raised'?: string;
+	'Depressed'?: string;
+	'Uniform'?: string;
+	'Drop shadow'?: string;
+	'Font Family'?: string;
+	'Proportional Sans-Serif'?: string;
+	'Monospace Sans-Serif'?: string;
+	'Proportional Serif'?: string;
+	'Monospace Serif'?: string;
+	'Casual'?: string;
+	'Script'?: string;
+	'Small Caps'?: string;
+	'Reset'?: string;
+	'restore all settings to the default values'?: string;
+	'Done'?: string;
+	'Caption Settings Dialog'?: string;
+	'Beginning of dialog window. Escape will cancel and close the window.'?: string;
+	'End of dialog window.'?: string;
+	'Seek to live, currently behind live'?: string;
+	'Seek to live, currently playing live'?: string;
+	'progress bar timing: currentTime={1} duration={2}'?: string;
+	'{1} is loading.'?: string;
+	'No content'?: string ;
+	'Color'?: string;
+	'Opacity'?: string;
+	'Text Background'?: string;
+	'Caption Area Background'?: string;
+	'Skip forward {1} seconds'?: string;
+	'Skip backward {1} seconds'?: string;
+	'Playing in Picture-in-Picture'?: string;
+}
```
We don't care about these so I think it should just be opaque:
```suggestion
export type VideoPlayerLocale = unknown;
```

RichardHelm commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,237 @@
+}
+
+/**
+ * Base class for video-player
+ *
+ * @public
```
```suggestion
 * @public
 * @component video-player
```

RichardHelm commented on libs/components/src/lib/video-player/video-player.ts:
```diff
@@ -0,0 +1,237 @@
+	/**
+	 * @internal
+	 */
+	player: any;
```
We can prefix it at least:
```suggestion
	_player: any;
```

TaylorJ76 replied:
We seem to have conflicting views on code conventions as Yonatan requested the prefix should be removed.

RichardHelm commented on libs/components/src/lib/video-player/video-player.spec.ts:
```diff
@@ -0,0 +1,389 @@
+	});
+
+	function menuItemsMatchReverseOrderOfRates(playbackRates: string[], playbackRatesMenuItems: NodeListOf<Element>) {
+		return playbackRates.reverse().reduce((acc, rate, index) => {
+			return acc &&
+				playbackRatesMenuItems[index].querySelector('.vjs-menu-item-text')!.textContent === `${rate}x`;
+		}, true);
```
```suggestion
		return playbackRates.reverse().every((rate, index) =>
			playbackRatesMenuItems[index].querySelector('.vjs-menu-item-text')!.textContent === `${rate}x`;
		);
```

RichardHelm commented on libs/components/src/lib/video-player/README.md:
````diff
@@ -0,0 +1,180 @@
+
+```html preview
+<style>
+	html { max-inline-size: 800px }
````
How can consumers change the size of the video-player? Without changing <html>?
Also, the video player should not be display `inline`

TaylorJ76 replied:
The Video is fluid and expands to fill the size of it's container, so consumers have control based on where it is placed in the page.
